ddtrace/contrib/celery/signals.py

```diff
@@ -82,11 +88,16 @@ def trace_before_publish(*args, **kwargs):
     span.set_tag(c.TASK_TAG_KEY, c.TASK_APPLY_ASYNC)
     span.set_tag('celery.id', task_id)
     span.set_tags(tags_from_context(kwargs))
+
     # Note: adding tags from `traceback` or `state` calls will make an
     # API call to the backend for the properties so we should rely
     # only on the given `Context`
     attach_span(task, task_id, span, is_publish=True)
 
+    headers = kwargs.get('headers')
+    if config.celery['distributed_tracing'] and headers is not None:
```

Comment:
should we add headers to `kwargs` if it doesn't exist? what are the cases when headers won't be present?

can `headers` be set in `args`?

Reply:
I believe `headers` should always exist and be a dictionary in the current version of Celery, this is being cautious for previous versions. I'll attempt to verify that further when I get this installed in a running Celery app for testing.

Reply:
Should we inject `header` if it is `None`?

e.g.

```python
if config.celery['distributed_tracing']:
    kwargs.get('headers')
    if kwargs.get('headers'):
        kwargs['headers'] = dict()

    # inject into `headers` ....
```

If there ever was a case when `headers` is not provided, do we still want to attempt to distribute?


(also, totally down for being cautious! in general the approach is good)

Reply:
I'm trying to dig in more in Celery to understand exactly how all this works, but I'm not having much luck either with the documentation or the code itself.

I did find this note in the Task docs:

> headers: | Mapping of message headers sent with this task message (may be None).

I'm not sure that the headers being `None` has any special meaning, though, as you're definitely allowed to write to them during these signals. I think we should still be fine to distribute in that case.